apps/iam/pkg/reconcilers/folder_reconciler.go

```diff
@@ -0,0 +1,204 @@
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana-app-sdk/app"
+	"github.com/grafana/grafana-app-sdk/k8s"
+	"github.com/grafana/grafana-app-sdk/logging"
+	"github.com/grafana/grafana-app-sdk/operator"
+	"github.com/grafana/grafana-app-sdk/resource"
+	foldersKind "github.com/grafana/grafana/apps/folder/pkg/apis/folder/v1beta1"
+	"github.com/grafana/grafana/pkg/services/authz/zanzana"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/client-go/rest"
+)
+
+// FolderStore interface for retrieving folder information
+type FolderStore interface {
+	GetFolderParent(ctx context.Context, namespace, uid string) (string, error)
+}
+
+// PermissionStore interface for managing folder permissions
+type PermissionStore interface {
+	GetFolderParents(ctx context.Context, namespace, folderUID string) ([]string, error)
+	SetFolderParent(ctx context.Context, namespace, folderUID, parentUID string) error
+	DeleteFolder(ctx context.Context, namespace, folderUID string) error
+}
+
+// AppConfig represents the app-specific configuration
+type AppConfig struct {
+	ZanzanaAddr string
+}
+
+type FolderReconciler struct {
+	permissionStore PermissionStore
+	folderStore     FolderStore
+}
+
+func NewFolderReconciler(cfg app.Config) (operator.Reconciler, error) {
+	// Create patch client
+	patchClient, err := getPatchClient(cfg.KubeConfig, foldersKind.FolderKind())
+	if err != nil {
+		return nil, fmt.Errorf("unable to create patch client for FolderReconciler: %w", err)
+	}
+
+	// Extract Zanzana address from config
+	appCfg, ok := cfg.SpecificConfig.(AppConfig)
+	if !ok {
+		return nil, fmt.Errorf("invalid config type: expected AppConfig, got %T", cfg.SpecificConfig)
+	}
+
+	// Create Zanzana client
+	zanzanaClient, err := getZanzanaClient(appCfg.ZanzanaAddr)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create zanzana client: %w", err)
+	}
+
+	// Create dependencies
+	folderStore := NewAPIFolderStore(&cfg.KubeConfig)
+	permissionStore := NewZanzanaPermissionStore(zanzanaClient)
+
+	// Create the reconciler
+	reconciler, err := operator.NewOpinionatedReconciler(patchClient, "folder-iam-finalizer")
+	if err != nil {
+		return nil, err
+	}
+
+	folderReconciler := &FolderReconciler{
+		permissionStore: permissionStore,
+		folderStore:     folderStore,
+	}
+
+	reconciler.Reconciler = &operator.TypedReconciler[*foldersKind.Folder]{
+		ReconcileFunc: folderReconciler.reconcile,
+	}
+
+	return reconciler, nil
+}
+
+func getPatchClient(restConfig rest.Config, kind resource.Kind) (operator.PatchClient, error) {
+	clientGenerator := k8s.NewClientRegistry(restConfig, k8s.ClientConfig{})
+	return clientGenerator.ClientFor(kind)
+}
+
+func getZanzanaClient(addr string) (zanzana.Client, error) {
+	transportCredentials := insecure.NewCredentials()
+
+	dialOptions := []grpc.DialOption{
+		grpc.WithTransportCredentials(transportCredentials),
```

Comment:
We need `authnlib.NewTokenExchangeClient` here, otherwise zanzana won't authorize request. But maybe it's a bit more complex topic. Now we authorize grafana instances on per-stack basis, so every stack can only access it's own store in openFGA, this prevents unauthorized reads/writes. But since operator is multi-tenant, we need to think about it's own authorization.